numpy/_core/src/multiarray/einsum_sumprod.cpp

```diff
@@ -0,0 +1,2181 @@
+/*
+ * This file provides optimized sum of product implementations used internally
+ * by einsum.
+ *
+ * Copyright (c) 2011 by Mark Wiebe (mwwiebe@gmail.com)
+ * The University of British Columbia
+ *
+ * See LICENSE.txt for the license.
+ *
+ */
+
+#define NPY_NO_DEPRECATED_API NPY_API_VERSION
+#define _MULTIARRAYMODULE
+
+#include "einsum_sumprod.h"
+
+#include "common.h"
+#include "einsum_debug.h"
+
+#include "simd/simd.h"
+#include <array>
+#include <iostream>
+#include <numpy/arrayobject.h>
+#include <numpy/halffloat.h>
```

Comment:
```suggestion
#include "common.hpp"
```
Use the C++ class `Half` instead, which is implemented by `common/half.hpp`, rather than the C wrapper.